### PR TITLE
chore: simplify balance charts

### LINKED ISSUE #1874: App.shapeshift.com All time frame chat on dashboard showing incorrectly.
### Overview

When viewing the all-time frame on the dashboard for app.shapeshift.com, it seems like it is off by an order of magnitude.

https://discord.com/channels/554694662431178782/978713734950289438

### References and additional details

repo steps 
1) connect to demo wallet or any wallet
2) wait for the page to load and click all timeframe on the dashboard chart
3) Observe the chart being off significantly from the balance of the wallet

image
the 

### Acceptance Criteria

For the all chart on the dashboard to display the proper information.

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by 0xdef1cafe**
## Description

* there was cosmos specific asset logic stinking up the balance chart. this moves it out to the selector layer and generalizes it to support all assets capable of delegation, and includes delegation, redelegation, and pending redelegation balances.
* balance charts should be faster on initial load as we no longer debounce to guess when tx history is loaded - we now know as we fetch historical tx's restfully and set a loaded state on completion
* this cleanup gets the balance charts in a better place to be turned into rainbows.

TODO

* [ ] figure out cosmos residual balance - **not a regression on this PR**

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [x] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #1874 

## Risk

this should actually fix the disparity between the right hand side of the charts and the total portfolio balance displayed. they still use different logic to calculate the same total but this can be unified in an additional PR.

## Testing

* a general regression of balance charts
* **please test with native wallet including new UTXO coins (DOGE/BCH/LTC)** - these previous would have gone to negative balances
* specific regression test of accounts with delegated, pending undelegations, and redelegations would be most useful

## Screenshots (if applicable)

* no specific user facing changes, charts aesthetically should look the same but
  * show correct data including all delegations, pending undelegations, and redelegations.
  * not go negative in the past for UTXO coins

### CODE REVIEW

File: src/state/slices/portfolioSlice/selectors.ts [hunk 1]
```diff
@@ -364,25 +365,17 @@ export const selectPortfolioCryptoHumanBalanceByFilter = createSelector(
 export const selectStakingDataByAccountSpecifier = createSelector(
   selectPortfolioAccounts,
   selectAccountSpecifierParamFromFilter,
-  (portfolioAccounts, accountSpecifier) => {
+  (portfolioAccounts, accountSpecifier): StakingDataByValidatorId | null => {
     return portfolioAccounts?.[accountSpecifier]?.stakingDataByValidatorId || null
   },
 )
 
-export const selectAllStakingDataByValidator = createSelector(
```
Comment (0xdef1cafe) [inline src/state/slices/portfolioSlice/selectors.ts:372]: this was literally the same as above with a different name 🤷

File: src/state/slices/portfolioSlice/selectors.ts [hunk 2]
```diff
@@ -459,24 +452,55 @@ export const selectTotalCryptoBalanceWithDelegations = createSelector(
...
+      balanceThreshold,
+      portfolioAccounts,
+      accountId,
+    ): PortfolioBalancesById => {
+      const rawBalances = (accountId ? accountBalances[accountId] : assetBalances) ?? {}
+      // includes delegation, redelegation, and undelegation balances
+      const totalBalancesIncludingAllDelegationStates: PortfolioBalancesById = Object.entries(
+        portfolioAccounts,
+      ).reduce((acc, [accountSpecifier, account]) => {
+        Object.entries(account?.stakingDataByValidatorId ?? {}).forEach(
```
Comment (0xdef1cafe) [inline src/state/slices/portfolioSlice/selectors.ts:487]: `stakingDataByValidatorId` generalizes for all CosmosSDK-based chains that support the concept of staking to validators

File: src/state/slices/portfolioSlice/selectors.ts [hunk 3]
```diff
@@ -459,24 +452,55 @@ export const selectTotalCryptoBalanceWithDelegations = createSelector(
...
+          ([_validatorPubKey, stakingDataByAccountSpecifier]) => {
+            const stakingData = stakingDataByAccountSpecifier[accountSpecifier]
+            const { delegations, redelegations, undelegations } = stakingData
+            const redelegationEntries = redelegations.flatMap(redelegation => redelegation.entries)
+            const combined = [...delegations, ...redelegationEntries, ...undelegations]
+            combined.forEach(entry => {
+              const { assetId, amount } = entry
+              acc[assetId] = bnOrZero(acc[assetId]).plus(amount).toString()
```
Comment (0xdef1cafe) [inline src/state/slices/portfolioSlice/selectors.ts:495]: any account with a staked balance in any state will by definition have a balance for the asset

File: src/hooks/useBalanceChartData/useBalanceChartData.ts [hunk 4]
```diff
@@ -359,26 +322,13 @@ export const useBalanceChartData: UseBalanceChartData = args => {
...
   const txFilter = useMemo(() => ({ assetIds, accountIds }), [assetIds, accountIds])
 
-  // we can't tell if txs are finished loading over the websocket, so
-  // debounce a bit before doing expensive computations
-  const txs = useDebounce(
-    useAppSelector(state => selectTxsByFilter(state, txFilter)),
-    500,
-  )
+  const txs = useAppSelector(state => selectTxsByFilter(state, txFilter))
+  const txHistoryStatus = useSelector(selectTxHistoryStatus)
```
Comment (0xdef1cafe) [inline src/hooks/useBalanceChartData/useBalanceChartData.ts:326]: consume our new tx loading status

File: src/hooks/useBalanceChartData/useBalanceChartData.ts [hunk 5]
```diff
@@ -315,41 +301,18 @@ type UseBalanceChartDataArgs = {
...
-    return acc
-  }, '')
-
-  const cosmosAccountSpecifier = account ? toAccountId({ chainId: cosmosChainId, account }) : ''
-
-  const delegationTotal = useAppSelector(state =>
-    selectTotalStakingDelegationCryptoByAccountSpecifier(state, {
-      accountSpecifier: cosmosAccountSpecifier,
-      assetId: cosmosAssetId,
-    }),
```
Comment (0xdef1cafe) [inline src/hooks/useBalanceChartData/useBalanceChartData.ts:352]: this was all asset specific stuff stinking up the chart logic

File: src/hooks/useBalanceChartData/useBalanceChartData.ts [hunk 6]
```diff
@@ -235,10 +222,9 @@ export const calculateBucketPrices: CalculateBucketPrices = args => {
     // if we have txs in this bucket, adjust the crypto balance in each bucket
     txs.forEach(tx => {
       if (tx.fee && assetIds.includes(tx.fee.assetId)) {
-        // balance history being built in descending order, so fee means we had more before
-        // TODO(0xdef1cafe): this is awful but gets us out of trouble
-        // NOTE: related to utxo balance tracking, just ignoring bitcoin for now as our only utxo chain support
-        if (tx.chain !== KnownChainIds.BitcoinMainnet) {
+        // don't count fees for UTXO chains
+        if (fromChainId(tx.chain).chainNamespace !== CHAIN_NAMESPACE.Bitcoin) {
```
Comment (0xdef1cafe) [inline src/hooks/useBalanceChartData/useBalanceChartData.ts:226]: charts would have been wrong for DOGE/LTC/BCH but we didn't notice

File: src/hooks/useBalanceChartData/useBalanceChartData.ts [hunk 7]
```diff
@@ -27,22 +26,21 @@ import {
...
+  selectRebasesByFilter,
+  selectTxHistoryStatus,
   selectTxsByFilter,
 } from 'state/slices/selectors'
-import { selectRebasesByFilter } from 'state/slices/txHistorySlice/selectors'
 import { Tx } from 'state/slices/txHistorySlice/txHistorySlice'
 import { useAppSelector } from 'state/store'
 
-import { includeStakedBalance, includeTransaction } from './cosmosUtils'
+import { includeTransaction } from './cosmosUtils'
```
Comment (0xdef1cafe) [inline src/hooks/useBalanceChartData/useBalanceChartData.ts:43]: we still need this utility and can't handle this at the selector layer, as we need to consider the *fees* for these transactions, but not act on adjusting the *balances* due to these transactions.

File: src/state/slices/portfolioSlice/selectors.ts [hunk 8]
```diff
@@ -459,24 +453,58 @@ export const selectTotalCryptoBalanceWithDelegations = createSelector(
...
+      balanceThreshold,
+      portfolioAccounts,
+      accountId,
+    ): PortfolioBalancesById => {
+      const rawBalances = (accountId ? accountBalances[accountId] : assetBalances) ?? {}
+      // includes delegation, redelegation, and undelegation balances
+      const totalBalancesIncludingAllDelegationStates: PortfolioBalancesById = Object.values(
+        portfolioAccounts,
+      ).reduce((acc, account) => {
+        Object.values(account?.stakingDataByValidatorId ?? {}).forEach(
```
Comment (0xdef1cafe) [inline src/state/slices/portfolioSlice/selectors.ts:488]: this generalizes for all assets that support the concept of staking that are cosmossdk-based

File: src/hooks/useBalanceChartData/useBalanceChartData.ts [hunk 9]
```diff
@@ -315,41 +301,18 @@ type UseBalanceChartDataArgs = {
...
-
-  const cosmosAccountSpecifier = account ? toAccountId({ chainId: cosmosChainId, account }) : ''
-
-  const delegationTotal = useAppSelector(state =>
-    selectTotalStakingDelegationCryptoByAccountSpecifier(state, {
-      accountSpecifier: cosmosAccountSpecifier,
-      assetId: cosmosAssetId,
-    }),
+  const balances = useAppSelector(state =>
+    selectBalanceChartCryptoBalancesByAccountIdAboveThreshold(state, accountId),
```
Comment (0xdef1cafe) [inline src/hooks/useBalanceChartData/useBalanceChartData.ts:315]: new selector hides removed logic

File: src/hooks/useBalanceChartData/cosmosUtils.ts [hunk 10]
```diff
@@ -1,32 +1,5 @@
...
 
-import { Bucket } from './useBalanceChartData'
-
-const cosmosAssetId = toAssetId({
-  chainId: cosmosChainId,
-  assetNamespace: 'slip44',
-  assetReference: ASSET_REFERENCE.Cosmos,
-})
-
 export const includeTransaction = (tx: Tx): boolean =>
```
Comment (0xdef1cafe) [inline src/hooks/useBalanceChartData/cosmosUtils.ts:3]: we still need this, and i suspect we're missing a method here causing residual balances for cosmossdk based assets

Comment (0xdef1cafe) [reply src/hooks/useBalanceChartData/cosmosUtils.ts:3]: @gomesalexandre any ideas?

File: src/hooks/useBalanceChartData/useBalanceChartData.ts (hunk 6 above)
Comment (gomesalexandre) [reply src/hooks/useBalanceChartData/useBalanceChartData.ts:226]: You bet we did https://github.com/shapeshift/web/pull/2309#pullrequestreview-1060584975

File: src/hooks/useBalanceChartData/useBalanceChartData.ts (hunk 7 above)
Comment (gomesalexandre) [reply src/hooks/useBalanceChartData/useBalanceChartData.ts:43]: While at it, can we do a bit of renaming here?
e.g `shouldIncludeTransaction` on the consumption.
Not sure about the naming of the util itself, I would probably invert all the logical checks and call it `isCosmosTransaction`

File: src/hooks/useBalanceChartData/cosmosUtils.ts (hunk 10 above)
Comment (gomesalexandre) [reply src/hooks/useBalanceChartData/cosmosUtils.ts:3]: Let's sync on this one with @elmutt, this line gives a bit of context already:
https://github.com/shapeshift/web/blob/f6eee96660e6f875186779f15ca370dfa17dd996/src/hooks/useBalanceChartData/useBalanceChartData.ts#L248

tl;dr from what I remember reviewing this, and reading the PR https://github.com/shapeshift/web/pull/1354 again is un/delegate Txs have a value, and we don't want to account for the value itself on the balance charts, only the balance change

**Review by gomesalexandre — COMMENTED** [review]
Tested with an account with negative balances and UTXOs + ATOM/OSMO staking operations:

It seems to improve - not fix - the zero issue at the beginning of charts, see localhost vs. prod:

image
image

I think the remaining issue here is the `includeTransaction` doesn't properly exclude claim Txs, I pinged @elmutt to get more context. 

I didn't find any regressions on this PR, either on an account with UTXOs and ATOM/OSMO advanced feature Txs or an account with EVM chains Txs

File: src/hooks/useBalanceChartData/useBalanceChartData.ts (hunk 7 above)
Comment (gomesalexandre) [reply src/hooks/useBalanceChartData/useBalanceChartData.ts]: Done in 96e442d36c0458cd10e81152e10a5c1a77cf3ae2 and cce5f1df78a75f3b8603404a3f088092ae8b267d

**Review by gomesalexandre — APPROVED** [review]
Approving as tested while contributing to this
